### PR TITLE
Adding an alternate layout parameter to custom fields

### PR DESCRIPTION
**Pull request by Bakual**
Pull Request for Issue #18162 .

### Summary of Changes
Adds a new parameter "Layout" to custom fields which allows to select a different field layout for the automatic display.

### Testing Instructions
* Create alternative layouts for a custom field. The original one is `components/com_fields/layout/field/render.php`, it can be overriden at component and template level, so for com_content in `components/com_content/layout/field/`, `templates/your_template/html/layouts/com_content/field` and `templates/your_template/html/layouts/com_fields/field`. The file isn't allowed to contain an underscore.
* Create a custom field and change the layout for that field, assign some value to it
* Check in frontend if that layout is used as expected.

### Expected result
Layout is used as specified

### Actual result
Not possible to choose a different layout

### Documentation Changes Required
Since it is a new feature, it has to be documented.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_fields.ini [hunk 1]
```diff
@@ -22,6 +22,8 @@ COM_FIELDS_FIELD_DISPLAY_LABEL="Automatic Display"
 COM_FIELDS_FIELD_DISPLAY_NO_DISPLAY="Do not automatically display"
 COM_FIELDS_FIELD_GROUP_DESC="The group this field belongs to."
 COM_FIELDS_FIELD_GROUP_LABEL="Field Group"
+COM_FIELDS_FIELD_LAYOUT_DESC="Choose an alternate layout to be used with automatic display"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_fields.ini]: It is a sentence please add a . at the end

Comment (Bakual) [reply administrator/language/en-GB/en-GB.com_fields.ini]: Sure thing!

**Comment by Bakual** [comment]
One thing I noted is that when you choose an override from a template which isn't active, the field will not show. JLayouts only work from the active template.

File: administrator/manifests/packages/pkg_weblinks.xml [hunk 2]
```diff
@@ -1,30 +0,0 @@
-<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (laoneo) [inline administrator/manifests/packages/pkg_weblinks.xml]: Is this removal made on purpose in this pr?

Comment (Bakual) [reply administrator/manifests/packages/pkg_weblinks.xml]: Oops, no. That one sneaked it.

**Comment by AndySDH** [comment]
Hey there,
great job with doing this!

I'm testing it and it does work correctly :) There are a couple of issues to address though:

- I noticed that you've put a "check" that hides the select input if Automatic Display is set to "No". However, you shouldn't do this, because this layout selection should (and indeed does) come into place also when you insert a field as part of a Field Group in the content with the content plugin, as described in #18162. So the select input field is always valid and doesn't need to be hidden away if Automatic Display is set to "No".

- I had my layouts in `templates/your_template/html/layouts/field` (which worked fine for the Content Plugin) and they were not showing up here in the select, so I think this folder is missing from the list. Yes, I know, yet another folder to add... didn't realize there were so many. Not sure why both this and `templates/your_template/html/layouts/com_fields/field` are used...

- Another thing, the "render" layout should probably not show up in the select list, as that equals to "Use Default". Unless by "Use Default" you mean the very default from the core which is not overriden? Because if you include an overridden render.php, then it should not be in the list. Maybe you could change the default option to "Use Default (render)"?

Also, a question, why is it that a filename cannot have an underscore? So far I've been using underscores for the filenames of my layouts manually used with the Content Plugin and they've been working normally. Not a big deal, just wondering.

Great job again!

**Comment by ghost** [comment]
@AndySDH  please mark your Test as successfully:

- open [Issue Tracker](https://issues.joomla.org/tracker/joomla-cms/18571)
- Login with your github-Account
- Click on blue "Test this"-Button above Authors-Picture
- mark your Test as successfully
- hit "submit test result"

**Comment by AndySDH** [comment]
@franz-wohlkoenig Sure, thanks for the guidelines! I'll be waiting for @Bakual about the fixes above and then will do that :)

**Comment by Bakual** [comment]
> because this layout selection should (and does come) into place also when you insert a field as part of a Field Group in the content with the content plugin

Ah true, didn't think of that. My though was to prevent misunderstanding when someones wants to insert a single field manually. I think then the layout doesn't apply (but haven't tested) and that could be confusing.

>  `templates/your_template/html/layouts/field`

That is then a lookup folder from JLayouts overriding mechanism itself. Looks indeed like I have to add that one as well. The com_fields one is because that is where the default layout actually is stored.

> Another thing, the "render" layout should probably not show up in the select list

Yes, it should not. The "Use Default" should cover that. That was my plan. Do you have an override for that in your template (I think I didn't test that case)? Or why did it show up?

> Also, a question, why is it that a filename cannot have an underscore?

Not really a technical reason. I based the formfield on the component layout one. In view layouts, we use underscore in layout files which should not appear in the select list. That allows to have "children" layouts. If you for example look at the article view, you see the `default.php` and a `default_link.php` file. The latter is a child layout of the former and doesn't appear in the article layout list.
I thought it makes sense to keep that naming convention to allow "hiding" certain layout files.

**Comment by AndySDH** [comment]
> Ah true, didn't think of that. My though was to prevent misunderstanding when someones wants to insert a single field manually. I think then the layout doesn't apply (but haven't tested) and that could be confusing.

You have a point. Yeah, the layout indeed does not apply when you insert a single field manually. Maybe for clarity we could rename the option to "Automatic Layout", and in the option description, change it to "Choose an alternate layout to be used with Automatic Display or as part of a Field Group."

Unless... we want to actually make it work also for when inserting a single field manually? (if that is even possible). This way, the option could act as "Default/Base Layout", which is always used for that field unless otherwise overridden in the single field via content plugin.

> That is then a lookup folder from JLayouts overriding mechanism itself. Looks indeed like I have to add that one as well. The com_fields one is because that is where the default layout actually is stored.

Yeah, figured so... One more folder to add, damn! Haha.

> Yes, it should not. The "Use Default" should cover that. That was my plan. Do you have an override for that in your template (I think I didn't test that case)? Or why did it show up?

Yeah, I have an override for render.php now in `templates/your_template/html/layouts/com_fields/field` and it's showing up in the list, it looks like.

> In view layouts, we use underscore in layout files which should not appear in the select list. I thought it makes sense to keep that naming convention to allow "hiding" certain layout files.

Gotcha, that makes sense to me too. Good idea/decision 👍

**Comment by Bakual** [comment]
> Yeah, the layout indeed does not apply when you insert a single field manually. 

I took another approach. The layout now applies also when you insert a single field :smile: 
You can still manually choose a different layout when inserting the field manually. You can even use the default one if you use "render" as layout option.

> One more folder to add

Done.

> Yeah, I have an override for render.php now in templates/your_template/html/layouts/com_fields/field and it's showing up in the list, it looks like.

Took care of them as well now.

**Comment by AndySDH** [comment]
@Bakual Fantastic! Saw all your latest commits, perfect stuff. Everything looks thorughly taken care of. I will test this all once again with the latest changes and officially mark as successful.

**Review by AndySDH — APPROVED** [review]

**Review by AndySDH — APPROVED** [review]

**Review by AndySDH — APPROVED** [review]

**Comment by AndySDH** [comment]
I have tested this item :white_check_mark: successfully on 2d7e71e221435ee1475382b2a73041926af585cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18571">issues.joomla.org/tracker/joomla-cms/18571</a>.</sub>

**Comment by AndySDH** [comment]
Can others test this? :)

File: administrator/components/com_fields/models/fields/fieldlayout.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+			// Add a grouped list
+			$html[] = JHtml::_(
+				'select.groupedlist', $groups, $this->name,
+				array('id' => $this->id, 'group.id' => 'id', 'list.attr' => $attr, 'list.select' => $selected)
+			);
+
+			return implode($html);
+		}
+		else
```
Comment (Quy) [inline administrator/components/com_fields/models/fields/fieldlayout.php]: Remove the `else` clause and just have `return '';`

Comment (Bakual) [reply administrator/components/com_fields/models/fields/fieldlayout.php]: That's what I get by copy-pasting from `JFormFieldComponentlayout` 😄

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on ea054e20a4a2f51ea925af9eb4b009177c405b18<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18571">issues.joomla.org/tracker/joomla-cms/18571</a>.</sub>

**Comment by ghost** [comment]
@AndySDH can you please retest?

**Comment by AndySDH** [comment]
I have tested this item :white_check_mark: successfully on ea054e20a4a2f51ea925af9eb4b009177c405b18<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18571">issues.joomla.org/tracker/joomla-cms/18571</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by laoneo** [comment]
Sorry for being late to the party, but com_contact needs to be adjusted too https://github.com/joomla/joomla-cms/blob/staging/components/com_contact/layouts/fields/render.php#L70.

**Comment by Bakual** [comment]
Thanks for the reminder @laoneo !

**Comment by AndySDH** [comment]
I have tested this item :white_check_mark: successfully on 8ba9ff904d9b396256a03107566de667c3aeb288<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18571">issues.joomla.org/tracker/joomla-cms/18571</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8ba9ff904d9b396256a03107566de667c3aeb288<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18571">issues.joomla.org/tracker/joomla-cms/18571</a>.</sub>

**Comment by zero-24** [comment]
@Bakual when you get a minute can you please update this pr and fix the merge conflicts?

**Comment by Bakual** [comment]
@zero-24 Rebased and fixed the conflicts

**Comment by AndySDH** [comment]
By the way, @laoneo, your comments from: #20243
> I would rather see a new function in core to define a default layout for the > field and provide different layouts for the same field type. For the list it would mean to add a new file > here like /plugins/fields/list/tmpl/list_ul.php and then provide these layouts as option in the field.
> What do you guys think?

> For the people who are requesting a layout chooser per field, It got merged today by #18571.

This PR actually wasn't specifically about what you mentioned over there. This PR here added the ability to add and choose a Custom Layout at a FIELD level (so before even knowing what field type it's going to render). (override for: "components/com_fields/layouts/fields/render.php")

What you were proposing is to add and choose an even deeper Custom Layout at a VALUE level for specific field types, where you would be able to specify a layout for individual values for fields that have multiple values (such as list, checkboxes, repeatable) (override for "plugins/fields/list/tmpl/list.php"). Which is not possible right now and was not included in this PR, just so you know.

**Comment by AndySDH** [comment]
@laoneo In preparation for the repeatable/subform field finalization (#22446), I think we should indeed also implement what you suggested and what I mentioned in the post above, so the ability to create an alternate layout to individual field values as well.

This would be useful to create for example a table layout for the repeatable/subform field type, or as you mentioned, a UL layout for a list field type, with: "/plugins/fields/list/tmpl/list_ul.php"

What we currently can do with what this PR added is the ability to set an override for the global render.php layout, which is great, but in addition to this, we also need an override for individual values of specific field types that include multiple values, such as list, radio, checkboxes, and repeatable/subform.

In the Field Editing it would look something like:

image

And of course, the dropdown there would list the layout files ONLY for that field type. So a list would only display files included in "/plugins/fields/list/tmpl/"

What do you think? Can you or @Bakual create a PR for this?

**Comment by Bakual** [comment]
@AndySDH Please don't post on closed/merged PRs. Nobody is going to read and answer that.
Open a new issue instead.

**Pull request merged**